File: leetcode/101-200/134-gas-station/gas_station.py
"""LeetCode #134: Gas Station — greedy single pass.

If total(gas) >= total(cost) a unique start exists: track a running tank; when it
goes negative, no station up to here can be the start, so restart from the next.
Return that start index, or -1.
"""


def can_complete(gas, cost):
    total = 0
    tank = 0
    start = 0
    for i in range(len(gas)):
        d = gas[i] - cost[i]
        total += d
        tank += d
        if tank < 0:
            start = i + 1
            tank = 0
    return start if total >= 0 else -1


def lcg(seed, n, mod):
    x = seed
    out = []
    for _ in range(n):
        x = (x * 1103515245 + 12345) % 2147483648
        out.append(x % mod)
    return out


def main():
    MOD = 1000000007
    sink = 0
    cases = [
        ([1, 2, 3, 4, 5], [3, 4, 5, 1, 2]),   # 3
        ([2, 3, 4], [3, 4, 3]),               # -1
        ([5], [4]),                           # 0
        ([3, 3, 4], [3, 4, 4]),               # -1
    ]
    for t in range(6):
        n = 8 + t * 3
        gas = lcg(t + 1, n, 20)
        cost = lcg(t + 100, n, 20)
        cases.append((gas, cost))
    for gas, cost in cases:
        r = can_complete(gas, cost)
        print(f"n={len(gas)}: {r}")
        sink = (sink * 1000003 + (r + 2)) % MOD
    print(f"sink: {sink}")


if __name__ == "__main__":
    main()

File: leetcode/101-200/134-gas-station/bench/gas_station.c
/* Benchmark harness for LeetCode #134 — Gas Station.
 * Mirrors gas_station.kara algorithm-for-algorithm.
 */

#include <stdio.h>

#define NP 8
#define N 200000
#define ITERS 1200

static long long gases[NP][N];
static long long costs[NP][N];

static long long can_complete(const long long *gas, const long long *cost, long long n) {
    long long total = 0;
    long long tank = 0;
    long long start = 0;
    for (long long i = 0; i < n; i++) {
        long long d = gas[i] - cost[i];
        total += d;
        tank += d;
        if (tank < 0) {
            start = i + 1;
            tank = 0;
        }
    }
    return total >= 0 ? start : -1;
}

static void lcg(long long seed, long long n, long long cap, long long *dst) {
    long long x = seed;
    for (long long k = 0; k < n; k++) {
        x = (x * 1103515245 + 12345) % 2147483648LL;
        dst[k] = (x / 65536) % cap;
    }
}

int main(void) {
    for (long long j = 0; j < NP; j++) {
        lcg(j + 1, N, 100, gases[j]);
        lcg(j + 100, N, (j % 2 == 0) ? 90 : 110, costs[j]);
    }

    long long sink = 0;
    for (long long it = 0; it < ITERS; it++) {
        long long idx = (it * 3) % NP;
        sink = (sink + can_complete(gases[idx], costs[idx], N)) % 1000000007LL;
    }
    printf("%lld\n", sink);
    return 0;
}
